feat: New Print Format Builder

netchampfaris opened the pull request:
Refactor of Print Format Builder has been long overdue. This PR tries to fix the culmination of all the frustrations I have had with the current PFB.

## Features
- Better drag-n-drop implementation
- Inline section and field labels editing
- Intuitive table column editing
- Custom page margins
- Custom font
- Page breaks after sections
- Uses [weasyprint](https://doc.courtbouillon.org/weasyprint/stable/) as the PDF renderer

## Why weasyprint?

wkhtmltopdf uses a really old version of WebKit which doesn't even support standard flexbox implementation. This is the primary reason why I decided to look for an alternative. wkhtmltopdf's [status page](https://wkhtmltopdf.org/status.html) basically says that there is no promise that it will be updated to the latest WebKit engine.

However, weasyprint does not only fix the flexbox problem, it implements standard W3C CSS specifications, which means there is a standard feature to add page numbers in a PDF rather than a [JS hack](https://github.com/frappe/frappe/blob/develop/frappe/templates/print_formats/pdf_header_footer.html#L44-L58). Not only this, there is whole bunch of [standard CSS features](https://doc.courtbouillon.org/weasyprint/stable/api_reference.html#css) that allow you to customize a PDF. This was an exciting discovery.

## Demo

The new Print Format Builder works alongside the current one with a feature flag in the Print Format document. So it does not break existing formats and is opt-in for now.

<details>

<summary>Show GIF</summary>

![new-print-format-builder-2](https://user-images.githubusercontent.com/9355208/132291641-953aea08-6730-4cd9-97f2-1e930256fca2.gif)

</details>


## Installation

weasyprint is a pure python package but it depends on a few libraries. Follow the installation instructions at [weasyprint installation page](https://doc.courtbouillon.org/weasyprint/stable/first_steps.html#installation)

For macOS users, simple run:
```
brew install pango libffi
```

## ToDo

- [x] Fix print preview in form
- [x] Letterhead editing
- [x] Handle custom formats, for e.g., template for Sales Taxes and Charges

> no-docs <will-be-added-later>

netchampfaris committed:
feat: New Print Format Builder

- Print Format Builder Beta page
- Add margin fields in Print Format
- Using vuedraggable for drag and drop

netchampfaris committed:
fix: poor man's hot reload

netchampfaris committed:
fix: More features

- Show selection dialog if Print Format not selected
- Field component
- Common store
- Edit Field label inline
- Configure table columns
- Edit Custom HTML
- Preview

netchampfaris committed:
fix: Set table column width in percentages

netchampfaris committed:
fix: Add section page break

netchampfaris committed:
fix: Add inline preview

netchampfaris committed:
feat: Use weasyprint to generate PDF

- /printpreview route to preview HTML template

netchampfaris committed:
fix: Handle on_page_show

netchampfaris committed:
fix: add cairocffi as requirement

netchampfaris committed:
fix: Ability to set font via Google Font

netchampfaris committed:
fix: misc

- show "Not Saved" when dirty
- Toggle Preview only when format is saved
- save a new format on first load

sagarvora commented:
I've used weasyprint with Frappe before.. it's slower.

netchampfaris committed:
feat: More features

- Letterhead editing
- Edit Header and Footer
- Margin Text
- PrintFormatGenerator class handles generation of HTML and PDF and repeating of Header/Footer
- Simplify /printpreview
- Separate renderer files for each fieldtype

netchampfaris committed:
fix: field container height to avoid scroll

netchampfaris committed:
feat: Better Letterhead editing

- Image resize and align

surajshetty3416 commented:
@netchampfaris What's the update on this? Is this ready for QA testing for basic/core functionalities?

netchampfaris committed:
fix: Set route_options from links

netchampfaris committed:
fix: Show print preview on print page

asoral commented:
I like this interesting enhancement, can we do something like a birt report builder. It would be great if we can use sql query to get data from a table

netchampfaris committed:
fix: Page number

- simpler configuration
- remove arbitrary margin text configuration

netchampfaris committed:
feat: Field Templates

- standard field templates for rendering custom templates for fields for e.g., taxes in invoice
- ability to create user defined reusable templates that show up in print format builder

netchampfaris committed:
fix: customize button in print

netchampfaris committed:
fix: minor ui fixes

netchampfaris committed:
feat: Configurable font size

netchampfaris committed:
feat: Spacer and Divider components

netchampfaris committed:
fix: print style

- bordered table
- pdf column rendering fix
- document-header class

netchampfaris committed:
fix: add header in default layout

netchampfaris committed:
fix: section fields orientation

- Option to render fields in a section in left-right orientation
- this is useful in cases where rendering currency values one below the other makes more sense

netchampfaris committed:
fix: controls styling

netchampfaris committed:
feat: Google font selection

google_fonts.json with a decent selection of fonts

netchampfaris committed:
Merge branch 'develop' of https://github.com/frappe/frappe into print-format-builder-beta

netchampfaris commented:
@surajshetty3416 Should be ready to merge now. Will you do a final round of testing before we merge this?

netchampfaris committed:
fix: Use icons instead of labelled buttons for consistency

netchampfaris committed:
style: formatting

netchampfaris committed:
fix: fields container height

netchampfaris committed:
fix: Include print style and print format css

netchampfaris commented:
I would like to defer docs and test cases until we are out of beta and the current print format builder is deprecated. This is not finished yet, there is long trail of fixes incoming after it is in the hands of users.

gavindsouza commented on frappe/printing/page/print_format_builder_beta/print_format_builder_beta.py:
```diff
@@ -0,0 +1,15 @@
+import frappe
+
+
+@frappe.whitelist()
```
This can be `frappe.lru_cache()` instead. No need for Redis here

gavindsouza commented on frappe/utils/weasyprint.py:
```diff
@@ -0,0 +1,247 @@
+# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and Contributors
+# MIT License. See license.txt
```
```suggestion
# MIT License. See LICENSE
```

gavindsouza commented:
> I would like to defer docs and test cases until we are out of beta and the current print format builder is deprecated. This is not finished yet, there is long trail of fixes incoming after it is in the hands of users.

Fair enough but I think we should add barebones docs. Test cases could be deferred but, I have a feeling there should be some basic ones for the `weasyprint` APIs at this stage. Although, if you don't see the need for them, idm as long as they get added in the near future 😛

netchampfaris committed:
fix: Revert back to text buttons

surajshetty3416 reviewed (changes requested):
@netchampfaris Nice enhancements overall 👍 

Noticed a few things that can be fixed.

- ![Screenshot 2021-10-21 at 6 43 32 PM](https://user-images.githubusercontent.com/13928957/138287065-a602720b-fbe1-4a59-9265-e4becc8f4424.png)
	These icons are confusing because in form view it signifies "Print" and "Refresh"… most people will not try to click on these for “Preview” and “Reset". **Suggestion:** Add text to the button


- ![Screenshot 2021-10-21 at 6 42 54 PM](https://user-images.githubusercontent.com/13928957/138287134-83e003f8-20fc-46b8-8d95-1af031d811df.png)
	Background color of controls are not the same


- ![Screenshot 2021-10-21 at 6 40 55 PM](https://user-images.githubusercontent.com/13928957/138287164-d2e482c0-3a3f-4bfe-a326-283b09afef70.png)
	This!!


- ![Editing User New Print Format](https://user-images.githubusercontent.com/13928957/138287191-45b47ed0-3912-492a-b471-7050cb3ef33a.png)
	Changing these options should update the preview in real-time (at least for HTML preview). The preview feature shines while tweaking such variables. (We can skip this if this requires major changes or if there are performance concerns)

surajshetty3416 commented:
> Fair enough but I think we should add barebones docs. Test cases could be deferred but, I have a feeling there should be some basic ones for the `weasyprint` APIs at this stage. Although, if you don't see the need for them, idm as long as they get added in the near future 😛

Let's skip the documentation for now (to avoid docs for namesake). I have created a [backlog](https://frappe.io/app/backlog/BCK-LG-2021-0513) to ensure the documentation and tests are added by next month.

netchampfaris committed:
fix: Create letter head UX

netchampfaris committed:
fix: body height in html preview

netchampfaris committed:
fix: control styling in preview

surajshetty3416 committed:
fix: Update licence text

Co-authored-by: gavin <gavin18d@gmail.com>

surajshetty3416 committed:
refactor: Use LRU cache instead of redis cache

surajshetty3416 commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'develop' into print-format-builder-beta

surajshetty3416 commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'develop' into print-format-builder-beta

netchampfaris committed:
fix: lru_cache decorator usage

netchampfaris committed:
fix: Barcode rendering

- CSS variables in svg don't work in PDF
- styling is done via external css now to make it work in form as well as PDF

netchampfaris committed:
fix: Field rendering of every field
